src/mpid/ch4/src/ch4r_callbacks.h

```diff
@@ -114,14 +118,61 @@ static inline int MPIDI_handle_unexp_cmpl(MPIR_Request * rreq)
 
     if (MPIDI_CH4U_REQUEST(rreq, req->status) & MPIDI_CH4U_REQ_MATCHED) {
         match_req = (MPIR_Request *) MPIDI_CH4U_REQUEST(rreq, req->rreq.match_req);
+
+#ifndef MPIDI_CH4_DIRECT_NETMOD
+        if (unlikely(match_req && MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(match_req))) {
+            anysource_partner = MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(match_req);
+            if (!MPIR_STATUS_GET_CANCEL_BIT(anysource_partner->status)) {
+                mpi_errno = MPID_Cancel_recv(anysource_partner);
+                if (mpi_errno != MPI_SUCCESS) {
+                    goto fn_fail;
+                }
+                /* What should we do if the anysource partner request is not canceled? */
+                MPIR_Assertp(MPIR_STATUS_GET_CANCEL_BIT(anysource_partner->status));
+            }
+            MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(match_req) = NULL;
+            MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(anysource_partner) = NULL;
+        }
+#endif /* MPIDI_CH4_DIRECT_NETMOD */
+
     } else {
         /* MPIDI_CS_ENTER(); */
-        if (root_comm)
+        if (root_comm) {
+#ifdef MPIDI_CH4_DIRECT_NETMOD
             match_req =
                 MPIDI_CH4U_dequeue_posted(MPIDI_CH4U_REQUEST(rreq, rank),
                                           MPIDI_CH4U_REQUEST(rreq, tag),
                                           MPIDI_CH4U_REQUEST(rreq, context_id),
                                           &MPIDI_CH4U_COMM(root_comm, posted_list));
+#else /* MPIDI_CH4_DIRECT_NETMOD */
+            while (TRUE) {
+                match_req =
+                    MPIDI_CH4U_dequeue_posted(MPIDI_CH4U_REQUEST(rreq, rank),
+                                              MPIDI_CH4U_REQUEST(rreq, tag),
+                                              MPIDI_CH4U_REQUEST(rreq, context_id),
+                                              &MPIDI_CH4U_COMM(root_comm, posted_list));
+
+                if (match_req && MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(match_req)) {
+                    anysource_partner = MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(match_req);
+                    if (!MPIR_STATUS_GET_CANCEL_BIT(anysource_partner->status)) {
+                        mpi_errno = MPID_Cancel_recv(anysource_partner);
+                        if (mpi_errno != MPI_SUCCESS) {
+                            goto fn_fail;
+                        }
+                        if (!MPIR_STATUS_GET_CANCEL_BIT(anysource_partner->status)) {
+                            anysource_partner = NULL;
+                            MPIR_Comm_release(root_comm);       /* -1 for posted_list */
+                            continue;
+                        }
+                    }
+                    MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(match_req) = NULL;
+                    MPIDI_CH4I_REQUEST_ANYSOURCE_PARTNER(anysource_partner) = NULL;
+                }
+
+                break;
+            }
```

Comment:
> I do not understand the logic of this while loop. Does it dequeue every matched-posted-request, and try to cancel the partner-request if it exists, then if the cancel fails, it dequeue and try-cancel the next matched-post-request ? Can you explain more ?

Again, I'm working on improving the documentation here:

This while loop is to ensure that if the first request dequeued from the posted list  turns out to be an MPI_ANY_SOURCE requests which was already matched by the other transport (netmod or shared memory), we can keep looking for the next request in the  posted queue.

Reply:
> Is MPIR_Comm_release(root_comm); /* -1 for posted_list */ for the canceled anysource_partner ?

This is for this request's reference on the communicator. The partner will downref for itself when that request is finished.

Reply:
I am asking because I see two more downref at line 177. One is commented as for posted_list. Ah, is the above downref only performed for the request whose partner was not canceled successfully ? Maybe your documentation could help.